parsons/databases/discover_database.py

```diff
@@ -0,0 +1,35 @@
+import os
+
+from parsons.databases.database_connector import DatabaseConnector
+from parsons.databases.redshift import Redshift
+from parsons.databases.mysql import MySQL
+from parsons.databases.postgres import Postgres
+from parsons.google.google_bigquery import GoogleBigQuery
+
+
+def discover_database() -> DatabaseConnector:
+    """Create an appropriate ``DatabaseConnector`` based on environmental variables.
+
+    Will search the environmental variables for the proper credentials for the
+    Redshift, MySQL, Postgres, and BigQuery connectors. See the documentation
+    for the connectors to variables required to initialize them.
+
+    If no suitable configuration is found, will raise an error.
+
+    Note that the variables to be searched for are hard-coded in this function,
+    since they are unlikely to change. If that is done, for some reason, or a
+    new database connector is added, ``discover_database`` should be updated
+
+    Returns:
+        DatabaseConnector: The database connector configured in the environment.
+    """
+    if os.getenv("REDSHIFT_PASSWORD"):
```

Comment:
I am slightly concerned about the logic here, where if someone happens to have two different db environmental variables set and wants to use the DB connector that is later in this list, they'd get an unexpected result. But I imagine "someone happens to have two different db environmental variables set" is probably pretty unlikely in practice.

Reply:
Love the idea of this function! Would be great to parameterize what database the connection points to.

However, I think Shauna has a point and don't think we should do this globally. For example, a pipeline from a Postgres application database to a Redshift data warehouse is not all that uncommon in which case, the same env would need both Postgres and Redshift credentials.

I think in that type of situation, you'd want something like
```
app_db = fetch_application_database_connector()
data_warehouse = fetch_datawarehouse_connector()
```

Perhaps we just do a relatively simple `fetch_database_connector` function that accepts a database type argument (preferably an enum) and then forwards all the arguments. So my toy example above for a postgres-redshift pipeline would look something like:

```
app_db = fetch_database_connector(database_type=os.getenv('APPLICATION_DATABASE_TYPE'))
data_warehouse = fetch_database_connector(database_type=os.getenv('DATA_WAREHOUSE_TYPE'))
```

And by passing through the arguments, we could support multiple database connections of the same type. For instance, perhaps there's one _main_ Postgres database but someone has a second software application. They'd want to define the default Postgres credentials in their env but then have the ability to connect to the second db when necessary.

Then that would look like (pretending the second software application was called widget):
```
app_db = fetch_database_connector(
        database_type=os.getenv('WIDGET_DATABASE_TYPE'),
        username = os.environ.get("WIDGET_USERNAME"),
        password = os.environ.get("WIDGET_PASSWORD"),
        ...
)
data_warehousse = fetch_database_connector(database_type=os.getenv('DATA_WAREHOUSE_TYPE'))
```

Reply:
Those are valid concerns. I wanted to keep it simple from a user-standpoint, so I kept all of the environmental variable checking inside the function. However I did make a few changes:

- If multiple databases are detected, it raises an error _unless_
- You can specify a default database by simply passing in the class: `discover_database(default_connector=MySQL)`
- If multiple databases are detected but the default is not among them, it errors
- You can pass in an ordered list of preference like this, and it will default to the first one given in the list if it finds multiple, otherwise return the one that is found: `discover_database(default_connector=[MySQL, Redshift])`

The logic is a bit intricate so I added a test file `tests\test_databases\test_discover_database.py` with nine test cases.

Reply:
I also think that for some of these more specialized cases, like where you want to control exactly what username gets passed into what, the recommendation should be to simply instantiate the database directly. I'm not sure if something like this:

```
app_db = fetch_database_connector(
        database_type=os.getenv('WIDGET_DATABASE_TYPE'),
        username = os.environ.get("WIDGET_USERNAME"),
        password = os.environ.get("WIDGET_PASSWORD"),
        ...
)
data_warehousse = fetch_database_connector(database_type=os.getenv('DATA_WAREHOUSE_TYPE'))
```

is better than:

```
app_db = Postgres(...)
warehouse_db = Redshift(...)
```

To expound, I don't think that discover_database should be a recommendation for all-purpose usage. If someone is an engineer at a TMC member org and they have a PostgreSQL data warehouse, they should probably just use 

```
db: DatabaseConnector = Postgres()
# or
db: Postgres = Postgres()
```

instead of `discover_database`. I think the main benefit of having `discover_database` is:

1. Sample scripts and tutorials
2. Utility functions inside the Parsons library like this one https://github.com/move-coop/parsons/pull/841
3. Other libraries that might want to use Parsons to have portable data-warehouse routines.

Reply:
Thanks for the clarification, Jason. I was initially imagining this might get used in user scripts but I see your point now about it mostly being helpful for situations where you don't know what database connector might be found, like a sample script or an internal utility function. If you know what databases you have you can just instantiate them directly. (Although I don't think Parsons currently lets you instantiate two connectors of the same type to two different databases, because the environmental variables would override each other, which is it's own separate issue.)

@willyraedy does this resolve your question or are we misunderstanding the use cases you're talking about?

Reply:
I've thought about this more, and I think that what @willyraedy is suggesting is a different use case. The idea behind my `discover_database` function is that it's a super easy to use, no-thought function you can just drop into some code and it should do its best to just make the database connection part of the script work.

Willy, if I'm understanding you correctly, then you want to have a function that will automatically load the database connection from specific configuration for that purpose in the environment. This is kind of the opposite purpose of `discover_database`. It's a very advanced use case, where you want the ability to deploy the same piece of software in different environments and to be able to change the database that it's running against without any code changes. I think that's another really interesting utility function around `DatabaseConnector` and I can see why TMC would be interested in that.

I think it would be best to implement that as a separate function, because it has a different job.  Perhaps you could pass in a prefix for a set of environmental variables. The function gets the database type from PREFIX_DATABASE_TYPE. Then it gets all of the other env vars that start with that prefix and passes them (with the prefix stripped out) as keyword arguments to the connector class that it finds.

```python
app_db: DatabaseConnector = fetch_environment_connection(prefix="APP")
    # env: APP_DATABASE_TYPE = "Redshift"
    # env: APP_REDSHIFT_USERNAME = "..."
    # etc.

warehouse_db: DatabaseConnector = fetch_environment_connection(prefix="WAREHOUSE")
    # env: WAREHOUSE_DATABASE_TYPE = "GoogleBigQuery"
    # env: WAREHOUSE_GOOGLE_CREDENTIALS = "..."
    # etc.
```

This would allow you to hot-swap multiple databases being used in the code by just changing the environmental variables, which is definitely something `discover_database` does _not_ support.

Reply:
Something like this would work (although it would require that DatabaseConnectors only take string values, or at least can handle strings as any of their arguments):

```python
def fetch_environment_connection(prefix: str) -> DatabaseConnector:
    database_type = os.getenv(f'{prefix}_DATABASE_TYPE')

    if database_type is None:
        raise EnvironmentError(f"Missing environment variable: {prefix}_DATABASE_TYPE")

    try:
        database_class = globals()[database_type]
    except KeyError:
        raise EnvironmentError(f"Unsupported database type: {database_type}")

    if not issubclass(database_class, DatabaseConnector):
        raise EnvironmentError(f"{database_type} is not a subclass of DatabaseConnector")

    kwargs = {}
    for key, value in os.environ.items():
        if key.startswith(f'{prefix}_') and key != f'{prefix}_DATABASE_TYPE':
            key_without_prefix = key[len(f'{prefix}_'):]
            # Lower is required because env. vars. are always (often?) uppercased. If anyone
            # broke Python convention and used a capital letter in their constructor parameters
            # then this won't work. But don't do that >_>
            kwargs[key_without_prefix.lower()] = value

    instance = database_class(**kwargs)

    return instance
```